### ISSUE TITLE
Compatibility with the latest NumPy version 1.24 which removes "numpy.int" entry

### ISSUE THREAD

**Issue by CloudyDory** [fangwei123456/spikingjelly#401]
# **Read before creating a new issue**

- Users who want to use SpikingJelly should first be familiar with the usage of PyTorch. 
- If you do not know much about PyTorch, we recommend that the user can learn the [basic tutorials of PyTorch](https://pytorch.org/tutorials/).
- Do not ask for help with the basic conception of PyTorch/Machine Learning but not related to SpikingJelly. For these questions, please refer to Google or [PyTorch Forums](https://discuss.pytorch.org/).

**For faster response**

You can @ the corresponding developers for your issue. Here is the division:

| Features                         | Developers                                                   |
| -------------------------------- | ------------------------------------------------------------ |
| Neurons and Surrogate Functions  | [fangwei123456](https://github.com/fangwei123456)<br />[Yanqi-Chen](https://github.com/Yanqi-Chen) |
| CUDA Acceleration                | [fangwei123456](https://github.com/fangwei123456)<br />[Yanqi-Chen](https://github.com/Yanqi-Chen) |
| Reinforcement Learning           | [lucifer2859](https://github.com/lucifer2859)                |
| ANN to SNN Conversion            | [DingJianhao](https://github.com/DingJianhao)<br />[Lyu6PosHao](https://github.com/Lyu6PosHao) |
| Biological Learning (e.g., STDP) | [AllenYolk](https://github.com/AllenYolk)                    |
| Others                           | [Grasshlw](https://github.com/Grasshlw)<br />[lucifer2859](https://github.com/lucifer2859)<br />[AllenYolk](https://github.com/AllenYolk)<br />[Lyu6PosHao](https://github.com/Lyu6PosHao)<br />[DingJianhao](https://github.com/DingJianhao)<br />[Yanqi-Chen](https://github.com/Yanqi-Chen)<br />[fangwei123456](https://github.com/fangwei123456) |

We are glad to add new developers who are volunteering to help solve issues to the above table.

**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**SpikingJelly version**

`0.0.0.0.14`

**Description**
`numpy.int` was [deprecated in NumPy 1.20](https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations) and was [removed in NumPy 1.24](https://numpy.org/devdocs/release/1.24.0-notes.html#expired-deprecations). However, the latest spikingjelly stable version 0.0.0.0.14 still uses it in some of the code, and will cause error.

For example, in `spikingjelly\activation_based\auto_cuda\base.py` line 233-250,

```
for key, value in py_dict.items():
    ctype: str = self.cparams[key]
    if isinstance(value, torch.Tensor):
        if value.dtype == torch.float:
            assert startswiths(ctype, ('const float', 'float'))

        elif value.dtype == torch.half:
            assert startswiths(ctype, ('const half2', 'half2'))

    if isinstance(value, cupy.ndarray):
        if value.dtype == np.float32:
            assert startswiths(ctype, ('const float', 'float'))

        elif value.dtype == np.float16:
            assert startswiths(ctype, ('const half2', 'half2'))

        elif value.dtype == np.int:
            assert startswiths(ctype, ('const int', 'int'))
```

Hope this can be fixed soon.

**Comment by fangwei123456** [comment]
Hi, this problem has been fixed in the lastest master version. The released `0.0.0.0.14` version can not be modified. You can install SpikingJelly from the latest GitHub code to avoid this bug.

**Comment by fangwei123456** [comment]
@biphasic You can try to install from github source code (not from pip), and check if it is ok.

**Comment by biphasic** [comment]
yes I did that and it worked well! I benchmarked SpikingJelly against other frameworks, congrats to being the fastest :D https://open-neuromorphic.org/p/snn-library-benchmarks/

**Comment by AllenYolk** [comment]
✅ 已修复。master 版本在 `spikingjelly/activation_based/cuda_kernel/auto_cuda/base.py:340` 加了 `hasattr(np, "int")` 守卫：

```python
elif value.dtype == int or (
    hasattr(np, "int") and value.dtype == np.int
):
    assert startswiths(ctype, ("const int", "int"))
```

这样无论是 NumPy 1.20–1.23（保留 `np.int`）还是 NumPy ≥ 1.24（移除 `np.int`，需使用内置 `int`）都能正常运行。

⚠️ 这个修复仅在 master 分支，PyPI 上的 `0.0.0.0.14` 仍有问题。如需在 NumPy ≥ 1.24 下使用，请从 GitHub 安装：

```shell
git clone https://github.com/fangwei123456/spikingjelly.git
cd spikingjelly
pip install .
```

**Comment by AllenYolk** [comment]
✅ Fixed. Master adds a `hasattr(np, "int")` guard in `spikingjelly/activation_based/cuda_kernel/auto_cuda/base.py:340`:

```python
elif value.dtype == int or (
    hasattr(np, "int") and value.dtype == np.int
):
    assert startswiths(ctype, ("const int", "int"))
```

This works on both NumPy 1.20–1.23 (where `np.int` exists) and NumPy ≥ 1.24 (where `np.int` was removed and the builtin `int` is used).

⚠️ This fix is only on master. PyPI `0.0.0.0.14` still has the issue. To use SpikingJelly with NumPy ≥ 1.24, install from GitHub:

```shell
git clone https://github.com/fangwei123456/spikingjelly.git
cd spikingjelly
pip install .
```
